engines/supernova/game-manager.cpp

```diff
@@ -0,0 +1,1022 @@
+				_texts[r] = text[rq];
+				_sentenceNumber[r] = i;
+			}
+			sentence(i, false);
+		} else
+			rq += rowLength[i];
+	}
+
+	_currentSentence = -1;
+	do {
+		do {
+			updateEvents();
+			mousePosDialog(_mouseX, _mouseY);
+			g_system->updateScreen();
+			g_system->delayMillis(_vm->_delay);
+		} while (!_mouseClicked && !_vm->shouldQuit());
+	} while (_currentSentence == -1 && !_vm->shouldQuit());
+
+	_vm->renderBox(0, 138, 320, 62, kColorBlack);
+
+	if (number && _currentSentence != -1 && _texts[_rowsStart[_currentSentence]] != kStringDialogSeparator)
+		_currentRoom->removeSentence(_currentSentence, number);
+
+	_guiEnabled = true;
+	_vm->_allowLoadGame = true;
+
+	return _currentSentence;
+}
+
+void GameManager::mousePosDialog(int x, int y) {
+	int a = y < 141 ? -1 : _sentenceNumber[(y - 141) / 10];
+	if (a != _currentSentence) {
+		sentence(_currentSentence, false);
+		_currentSentence = a;
+		sentence(_currentSentence, true);
+	}
+}
+
+void GameManager::takeObject(Object &obj) {
+	if (obj.hasProperty(CARRIED))
+		return;
+
+	if (obj._section != 0)
+		_vm->renderImage(obj._section);
+	obj._click = obj._click2 = 255;
+	_inventory.add(obj);
+}
+
+void GameManager::drawCommandBox() {
+	for (int i = 0; i < ARRAYSIZE(_guiCommandButton); ++i) {
+		_vm->renderBox(_guiCommandButton[i]);
+		int space = (_guiCommandButton[i].width() - Screen::textWidth(_guiCommandButton[i].getText())) / 2;
+		_vm->renderText(_guiCommandButton[i].getText(),
+						_guiCommandButton[i].getTextPos().x + space,
+						_guiCommandButton[i].getTextPos().y,
+						_guiCommandButton[i].getTextColor());
+	}
+}
+
+void GameManager::drawInventory() {
+	for (int i = 0; i < ARRAYSIZE(_guiInventory); ++i) {
+		_vm->renderBox(_guiInventory[i]);
+		_vm->renderText(_inventory.get(i + _inventoryScroll)->_name,
+						_guiInventory[i].getTextPos().x,
+						_guiInventory[i].getTextPos().y,
+						_guiInventory[i].getTextColor());
+	}
+
+	_vm->renderBox(_guiInventoryArrow[0]);
+	_vm->renderBox(_guiInventoryArrow[1]);
+	if (_inventory.getSize() > ARRAYSIZE(_guiInventory)) {
+		if (_inventoryScroll != 0)
+			_vm->renderText(_guiInventoryArrow[0]);
+		if (_inventoryScroll + ARRAYSIZE(_guiInventory) < _inventory.getSize())
+			_vm->renderText(_guiInventoryArrow[1]);
+	}
+}
+
+int GameManager::getKeyInput() {
+	while (!_vm->shouldQuit()) {
+		updateEvents();
+		if (_keyPressed) {
+			return _key.ascii;
+		}
+		g_system->updateScreen();
+		g_system->delayMillis(_vm->_delay);
+	}
+	return 0;
+}
+
+Common::EventType GameManager::getMouseInput() {
+	while (!_vm->shouldQuit()) {
+		updateEvents();
+		if (_mouseClicked)
+			return _mouseClickType;
+		g_system->updateScreen();
+		g_system->delayMillis(_vm->_delay);
+	}
+	return Common::EVENT_INVALID;
+}
+
+void GameManager::getInput() {
+	while (!_vm->shouldQuit()) {
+		updateEvents();
+		if (_mouseClicked || _keyPressed)
+			break;
+		g_system->updateScreen();
+		g_system->delayMillis(_vm->_delay);
+	}
+}
```

Comment:
These three methods could be merged into one. You could add a parameter that would filter the expected event types, e.g. :

    int GameManager::getInput(bool checkMouseEvents, bool checkKeyboardEvents) {
        do {
            updateEvents();
            g_system->updateScreen();
            g_system->delayMillis(_vm->_delay);
        } while (!_vm->shouldQuit() && !(_mouseClicked && checkMouseEvents) && !(_keyPressed && checkKeyboardEvents));
    }